dali/test/python/test_operator_spectrogram.py

```diff
@@ -0,0 +1,227 @@
+import librosa as librosa
+import math
+
+class SpectrogramPipeline(Pipeline):
+    def __init__(self, device, batch_size, iterator, nfft, window_length, window_step,
+                 num_threads=1, device_id=0):
+        super(SpectrogramPipeline, self).__init__(batch_size, num_threads, device_id)
+        self.device = device
+        self.iterator = iterator
+        self.inputs = ops.ExternalSource()
+        self.fft = ops.Spectrogram(device = self.device,
+                                   nfft = nfft,
+                                   window_length = window_length,
+                                   window_step = window_step,
+                                   power = 2)
+
+    def define_graph(self):
+        self.data = self.inputs()
+        out = self.data.gpu() if self.device == 'gpu' else self.data
+        out = self.fft(out)
+        return out
+
+    def iter_setup(self):
+        data = self.iterator.next()
+        self.feed_input(self.data, data)
+
+def spectrogram_func_torch(nfft, win_len, win_step, input):
+    def hann_win(n):
+        hann = torch.ones([n], dtype=torch.float32)
+        a = (2.0 * math.pi / n)
+        for t in range(n):
+            phase = a * (t + 0.5)
+            hann[t] = 0.5 * (1.0 - math.cos(phase))
+        return hann
+
+    waveform = torch.FloatTensor(input)
+    assert nfft == win_len
+    spectrogram = torchaudio_transforms.Spectrogram(
+        n_fft=nfft, win_length=win_len, hop_length=win_step, power=2, window_fn=hann_win)
+    out = spectrogram.forward(waveform)
+    print(out)
+    return out
+
+
+def spectrogram_func_librosa(nfft, win_len, win_step, input_data):
+    def hann_win(n):
+        hann = np.ones([n], dtype=np.float32)
+        a = (2.0 * math.pi / n)
+        for t in range(n):
+            phase = a * (t + 0.5)
+            hann[t] = 0.5 * (1.0 - math.cos(phase))
+        return hann
+
+    has_channels_dim = len(input_data.shape) == 2
+
+    if has_channels_dim:
+        input_data = np.squeeze(input_data, axis=0)
+
+    out = np.abs(
+        librosa.stft(y=input_data, n_fft=nfft, hop_length=win_step, window=hann_win))**2
+
+    # Alternative way to calculate the spectrogram:
+    # out, _ = librosa.core.spectrum._spectrogram(
+    #     y=input_data, n_fft=nfft, hop_length=win_step, window=hann_win, power=2)
+
+    if has_channels_dim:
+        out = np.expand_dims(out, axis=0)
+
+    return out
+
+class SpectrogramPythonPipeline(Pipeline):
+    def __init__(self, device, batch_size, iterator, nfft, window_length, window_step,
+                 num_threads=1, device_id=0):
+        super(SpectrogramPythonPipeline, self).__init__(
+              batch_size, num_threads, device_id,
+              seed=12345, exec_async=False, exec_pipelined=False)
+        self.device = "cpu"
+        self.iterator = iterator
+        self.inputs = ops.ExternalSource()
+
+        function = partial(spectrogram_func_librosa, nfft, window_length, window_step)
```

Comment:
I would make this `spectrogram_func_librosa` an argument as you can use `spectrogram_func_torch` there as well.

Reply:
I removed spectrogram_func_torch, but I'll make it an argument anyway